File: src/main/java/leetcode/topinterview/LeetCode179.java
package leetcode.topinterview;

import java.util.Arrays;
import java.util.Comparator;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/27 13:23
 */
public class LeetCode179 {
    //copy
    //https://www.youtube.com/watch?v=LUxREjEADCw&t=215s
    public String largestNumber(int[] nums) {
        int n = nums.length;
        String[] numsToWord = new String[n];
        for (int i = 0; i < n; i++) {
            numsToWord[i] = String.valueOf(nums[i]);
        }
        // 7 8
        Arrays.sort(numsToWord, new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                //      87                  78
                return (o2 + o1).compareTo(o1 + o2);
            }
        });
        StringBuilder sb = new StringBuilder();

        for (String s : numsToWord) {
            sb.append(s);
        }
        String res = sb.toString();
        return res.charAt(0) == '0' ? "0" : res;
    }
}

File: src/main/GO/leetcode/lists/topinterview/LeetCode179.go
package topinterview

import (
	"sort"
	"strconv"
)

func largestNumber(nums []int) string {

	if len(nums) == 0 {
		return ""
	}
	list := make([]string, len(nums))
	for i := 0; i < len(nums); i++ {
		list[i] = strconv.Itoa(nums[i])
	}
	sort.Sort(strs(list))

	var res string

	for i := 0; i < len(list); i++ {
		res += list[i]
	}
	if res[0] == '0' {
		return "0"
	}
	return res
}

type strs []string

func (s strs) Len() int {
	return len(s)
}
func (s strs) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}
func (s strs) Less(i, j int) bool {
	return s[i]+s[j] > s[j]+s[i]
}
